### ISSUE TITLE
Windows issue

### ISSUE THREAD

**Issue by amitmerchant1990** [sindresorhus/emoj#5]
I've tried to use it on Windows 7 and it appears to be not working correctly. All I can see is below
![capture](https://cloud.githubusercontent.com/assets/3647841/16292659/24721f20-3931-11e6-996d-c2a687b0a36a.JPG)

Node.js version `4.4.5`.

**Comment by CodeDotJS** [comment]
If you are unable to read Unicode characters, it may be because your system is not properly configured.
I haven't used windows yet, but I think you need to change the font style of your console.

This issue has nothing to do with `node version` or `emoj`

**Comment by sindresorhus** [comment]
The default Windows terminal font doesn't support unicode and unlike macOS, the terminal doesn't fallback to other fonts that do have such character.

I would strongly recommend using [Bash for Windows](https://msdn.microsoft.com/en-us/commandline/wsl/about) instead. If it still doesn't display properly you might need to download a good fallback font. I'd recommend downloading the [Noto Color Emoji font](https://www.google.com/get/noto/#emoji-qaae-color) and setting it as your terminal font. Could you all try that?

**Comment by nolanlawson** [comment]
Bash for Windows doesn't show the emoji either. Explorer also complains that Noto Color Emoji is not a valid TTF file (https://github.com/googlei18n/noto-emoji/issues/43). Ah, Windows problems. :confused:

**Comment by sindresorhus** [comment]
@nolanlawson Yup: http://superuser.com/questions/1062418/how-do-i-replace-windows-10s-emoji

**Comment by jdalton** [comment]
\cc @bitcrazed.

**Comment by adueck** [comment]
Works ok-ish with Git Bash for Windows
https://git-scm.com/download/win

**Comment by sindresorhus** [comment]
@adueck Can you share a screenshot?

**Comment by aoozdemir** [comment]
I'm on windows too. I get this error when I try to run on git bash.

`git version 2.9.0.windows.1`, `npm --version 3.3.12`, `node --version v5.4.1`

```
C:\Users\...\AppData\Roaming\npm\node_modules\emoj\cli.js:36
process.stdin.setRawMode(true);
              ^

TypeError: process.stdin.setRawMode is not a function
    at Object.<anonymous> (C:\Users\...\AppData\Roaming\npm\node_modules\emoj\cli.js:36:15)
    at Module._compile (module.js:397:26)
    at Object.Module._extensions..js (module.js:404:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:429:10)
    at startup (node.js:139:18)
    at node.js:999:3
```

**Comment by bitcrazed** [comment]
Thanks for looping me in :) 

Running in Bash on Ubuntu on Windows in Windows 10 Insiders build 14374, I see this:
![image](https://cloud.githubusercontent.com/assets/961950/16364333/9517bc84-3b99-11e6-9252-5427c1042cf6.png)

Alas, this highlights the fact that emoji support is currently very limited by the Windows Console. We have plans to fix this in the future, but for now, alas we don't support emoji's sufficiently to make emoj work :(

This said, if you would like to see us improve our emoji support in the future, please do upvote some of these issues in our UserVoice: https://wpdev.uservoice.com/forums/266908-command-prompt-console/category/161892-bash?query=UTF-8

**Comment by adueck** [comment]
@aoozdemir Yes, Git Bash has some kind of problem working with with interactive node cli apps. When you install Git Bash it it tells you this and explains some other service that you can use to run those interactive cli apps. There is a way to do this, but I forgot what this is though and can't find it in the documentation. :cry: 

So I just run it with the argument in the initial command ie. `emoji 'text you want to use'` 

And it seems to work pretty well. Here's a screenshot.

<img width="251" alt="screenshot" src="https://cloud.githubusercontent.com/assets/12955501/16374804/40947320-3c69-11e6-9c96-b60311b57872.png">

**Comment by aoozdemir** [comment]
Ahh! Win terminal bothers me so much lately. ☺ I believe it's time for a brand new, beautiful terminal for Windows (especially for Win10).

**Comment by bitcrazed** [comment]
@aoozdemir Are you running the Win10 insiders builds? We've made A LOT of really big improvements to the Console, including big perf improvements, better High-DPI support, lots more VT sequence support (not done yet, but lots done):
![8-split-cacademo](https://cloud.githubusercontent.com/assets/961950/16390083/d8720838-3c55-11e6-9f5b-230d30d55699.png).

We are continuing to invest in improving our console experience. We have big plans for the future, but please bear-with us while we figure out how to deliver against the demands of so many customers ;)

**Comment by bitcrazed** [comment]
@adueck You're seeing some additional features that the Cygwin/MinGW team added to how they handle font-fallback.

Alas, we cannot add this mechanism to the built-in Windows console this time around, but we have several items on our backlog to fix a number of the issues that prevent us from being able to support many globalization/localization issues, including emoji!

**Comment by aoozdemir** [comment]
@bitcrazed no I didn't sign up for the insider program. That's actually great to hear, actually. I'm looking forward to try it out. :tada: 

(I'm not sure if it's on the table but, tabs would be so cool to have :smile:)

**Comment by bitcrazed** [comment]
Tabs ... and MANY, MANY other features ;)

**Comment by eosrei** [comment]
> Explorer also complains that Noto Color Emoji is not a valid TTF file

@nolanlawson It only contains CBLC/CBDT color glyphs. It doesn't have any b&w regular glyphs which makes it an invalid font. FontForge can't work with it either, it's not just a Windows problem. :)

My emoji fonts have SVGinOT color *and* b&w regular glyphs. They work fine on Windows, although only show color only in Firefox.
* https://github.com/eosrei/emojione-color-font
* https://github.com/eosrei/twemoji-color-font

@yincrash has created a method to make my EmojiOne-based font replace the Windows default emoji: https://github.com/eosrei/emjione-color-font/pull/47 I doubt it will solve this issue, but could help. Happy to answer any other questions.

**Comment by yincrash** [comment]
Unfortunately, Windows 7 which is mine and @amitmerchant1990's OS does not support Bash for Windows.

Additionally, [cmd.exe (which uses conhost.exe)](https://en.wikipedia.org/wiki/Win32_console) doesn't really support Unicode. It uses [code pages](https://msdn.microsoft.com/en-us/library/windows/desktop/dd317752(v=vs.85).aspx) :scream: (and UCS-2 under the hood). This means you have to change the code page whenever you want something outside your current code page ([437](https://en.wikipedia.org/wiki/Code_page_437) for Americans), and switching to the "UTF-8" (65001) codepage doesn't really work well, plus it for sure does not support anything outside the BMP such as most Emoji.

MSYS2 (and MINGW / Git Bash for Win) shell does support Unicode because they use Cygwin's [mintty](https://en.wikipedia.org/wiki/Mintty), because they aren't real Windows consoles (using conhost.exe), to normal windows (non-msys2/mingw packages), they are seen to apps like python and node as pipes and not as a console output. _However_, if you're using the MSYS2/MinGW version of python it does act like it's POSIX counterpart and outputs Emoji to the shell. I couldn't find a MSYS2 version of nodejs outside of 0.10.x version to test the `emoj` package.

Here's a screenshot of MSYS2 python displaying emoji when I set the console font to the B&W Noto Emoji. It doesn't seem that they use Window's fallback mechanisms for fonts, so you'd to merge a nice console font with an emoji font using FontTools or some other tool to get emoji working nicely.

![python-msys2-emoji](https://cloud.githubusercontent.com/assets/609567/16800700/d4bcb9ee-48b3-11e6-86cf-9aa5c0f2e28b.png)

tl;dr - windows console is a nightmare

**Comment by bitcrazed** [comment]
Thanks for the feedback. As the PM for Windows Console/ConHost and Bash experience, I appreciate your candor and can confirm that we are keenly aware of these and many other limitations that limit Console's value and cause frustration.

Rest assured that we are working hard to fix things and I think that you'll see a lot of under the covers improvements in Win10AU, when you do upgrade (please consider doing so if you can; not only is the free upgrade offer expiring soon, but you're missing out on a lot of really great improvements that'll make your experience much better).

**Comment by yincrash** [comment]
Oh I understand why it's like that, and legacy support is one of the biggest reasons I love Windows. It's really awesome the WSL is being developed to help include the Linux ecosystem better. I don't think the Win32 API designers originally expected us to use code points outside of UCS-2 very much in the console, and I don't fault them for that.

**Comment by bitcrazed** [comment]
@yincrash Yeah, NT implemented Unicode support VERY early in its development and in many areas (e.g. Console) has not done a good job of keeping up to date with improvements like UTF-8. We're aiming to fix that in future releases :)

**Comment by renatorib** [comment]
😁

**Comment by nnmrts** [comment]
I'm not a Windows user, but shouldn't this be fixed by now by the Windows "Terminal" that was released in 2019? Powershell should be able to display emoji too, I think.

I would also suggest changing the readme so windows users don't think they can't use this.

**Comment by HLFH** [comment]
> I'm not a Windows user, but shouldn't this be fixed by now by the Windows "Terminal" that was released in 2019? Powershell should be able to display emoji too, I think.

Nope. https://github.com/microsoft/terminal/issues/190

**Comment by bitcrazed** [comment]
@HLFH The issue you refer to is an issue in the legacy Windows Console, not the Windows Terminal. 

Windows Terminal is a new modern Terminal app in which we will implement all the features you and we want and need of a modern terminal app, including our new DirectX based text rendering engine that can display an ever increasing number of emoji and unicode glyphs:

![image](https://user-images.githubusercontent.com/961950/84070299-c426b300-a980-11ea-8660-1b23a27a8f6b.png)

However, Windows Console & Cmd shell's primary job is to preserve backward compatibility with legacy scripts, apps, and tools. As such, we avoid changing Console (and especially Cmd) unless we have to. When we do have to make a change, we have to take enormous care not to break anything - and you wouldn't believe the things that break when we make the smallest change to Console/Cmd!

This said, one of the key goals of implementing Terminal in native C++ was our ability to utilize and share code between Console and Terminal. We do have a goal of introducing Terminal's new text handling and text layout/rendering engines, along with the supporting technologies inc. our accessibility engine once sufficiently complete, stable, and performant enough. However, as you can imagine, this will take a little time.

This said, we do intend to introduce Terminal's text layout and rendering engine to Console which will enable it to display emoji and most unicode glyphs. However, before we can do that, we have to do the hard work. Stay tuned.

**Comment by nnmrts** [comment]
I'd really recommend avoiding cmd.exe as the main shell as a developer in 2020 though. Seriously, do yourself a favour and use a modern terminal emulator like the Windows Terminal, Powershell, hyper, etc... any of those.

Don't expect Emoji support from a decades old application that just got Unicode support as recently as October 2018. Don't expect anything from it other than itself "just working". Because that's basically the only purpose of it today.

**Comment by hackdefendr** [comment]
Windows Terminal latest version works close enough without special fonts. Works even better with the correct fonts.

**Comment by charlesroper** [comment]
Works great in Windows Terminal with PowerShell.

![emoj-on-windows](https://user-images.githubusercontent.com/176/151471274-3f3a7be9-c824-4fcd-b150-241ea9267b0c.gif)

**Comment by bitcrazed** [comment]
@nnmrts A few clarifications:

> I'd really recommend avoiding cmd.exe as the main shell as a developer in 2020 though. 

Couldn't agree more - Windows developers should be using PowerShell these days 😜

https://twitter.com/bitcrazed/status/1387520440529080321?s=20&t=DQspE6JWijN3OJpjuJpKbg
![image](https://user-images.githubusercontent.com/961950/151605745-10e26fd1-805b-4b6c-823f-6cc00ba147d9.png)


> Seriously, do yourself a favour and use a modern terminal emulator like the Windows Terminal, Powershell, hyper, etc... any of those.

Cmd and PowerShell are command-line shells and operate independently from the terminal used to command them.

Windows Terminal (and the legacy Console/ConHost) are terminals.

Many might find this series of blogs I wrote useful/interesting: https://devblogs.microsoft.com/commandline/windows-command-line-backgrounder/ 

> Don't expect Emoji support from a decades old application that just got Unicode support as recently as October 2018. 

Windows NT, upon which modern Windows is based, implemented 16-bit [UCS-2](https://en.wikipedia.org/wiki/Universal_Coded_Character_Set) back in 1989 before there even was a Unicode standard, and long before [UTF-8](https://datatracker.ietf.org/doc/html/rfc2279) was created.

> Don't expect anything from it other than itself "just working". Because that's basically the only purpose of it today.

As [Charles points out](https://github.com/sindresorhus/emoj/issues/5#issuecomment-1023802555), emoji works well on Windows Terminal 😁

**Comment by KMZ0209** [comment]
I really like your npm !! :)
"Works best on macOS and Linux in [readme.md](https://github.com/sindresorhus/emoj/blob/main/readme.md)." However, there are some emojis that do not output in the WSL2 Ubuntu 20.04 environment, so please fix them if you like. I would be happy if you could add that some emojis are not output on Ubuntu 20.04. :) Thanks.
![スクリーンショット-2023-09-30-151511](https://github.com/sindresorhus/emoj/assets/111144394/e36dbabb-da02-4352-954d-38ac3f9404f5)

**Comment by dnjondimackal-coder** [comment]
This works fine if Windows Terminal is used as the emulator instead of Windows Console Host
